### ISSUE TITLE
Slow training on custom dataset with a several validation datasets

### ISSUE THREAD

**Issue by shayan-nikoo** [HuguesTHOMAS/KPConv-PyTorch#225]
Hi Thomas,

I trained the Kpconv model by modifying `S3DIS.py`. 

I have 25 classes that I trained on 20 buildings each of size of 150mb-200mb with approximately 10 million points per file and in total of 4 GB. My data DOES NOT have RGB. I leave 4 buildings for test/validation.

The first try, I missed fixing the below part in `trainer.py` thus the evaluation was skipped and `val_preds` files were not generated in results. However, the training was done successfully in about **11 hours** and the results were good (attached image). 

**FIRST TRY (no validation- quick training)**
[https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/master/datasets/S3DIS.py](url)
```python
self.cloud_names = [filename for filename in os.listdir(self.path) if filename.startswith("Building")]  # 19 files in total
self.all_splits = list(range(0, len(self.cloud_names))) 
self.validation_split = [1, 3, 5, 7] # always should be a list even for one number.
```
I missed changing this part in `trainer.py`, so no validation:
[https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/680296878d238e6bdb798c190120062a46f492d1/utils/trainer.py#L431C1-L433C19](url)
```python
      # Do not validate if dataset has no validation cloud
      if val_loader.dataset.validation_split not in val_loader.dataset.all_splits:
          return
```
**Second TRY (validation on 4 files- 5x slower training)**
Then I fixed the above line in `trainer.py` as below according to the `validation_split` of `S3DIS.py` file and I'm getting validation results per epoch but the **training time now is about 5 times more**. It can take more than two days. The model doesn't use more than 7 GB of GPU and 16 GB of CPU while I have more processing power.

*Fixed in `trainer.py`*
```python
      # Do not validate if dataset has no validation cloud
      #if val_loader.dataset.validation_split not in val_loader.dataset.all_splits:
       #   return
        if not (set(val_loader.dataset.validation_split)).issubset(set(val_loader.dataset.all_splits)):
            Warning.warn(f"Validation set {val_loader.dataset.validation_split} is not in all_splits!!!")
            return
```

I'm not changing other parameters, keeping them as the original config as S3DIS.
This is the result of the First TRY. On the 2nd try I stopped as it is going very slow.
```
Confusion on sub clouds
50.85 | 96.87 95.17 75.43 71.26  0.00  0.00 54.71 39.19 53.93 63.72 61.84 88.24 43.18 31.75 18.78 61.48 49.26 74.97 28.13 27.75 66.11 75.08 37.20  6.28 50.85
```
<img width="1005" alt="kpconv_results" src="https://github.com/HuguesTHOMAS/KPConv-PyTorch/assets/123162959/97465f9d-be97-462c-97df-f926f2f6f7bd">

Q1. Any idea what am I missing, why the training time is much longer with validation? If I do validation only on one file is not it imbalance train/test sets?

Q2. How can I speed up training and inference by using more resources (I have 16 gb GPU  and 64 gb CPU RAM)? I know I can increase `first_subsampling_dl` and `in_radius` but that's a tradeoff with final accuracy and it needs trial and error. What about increasing the `batch_num = 6` ? or any better idea? 

Q3. What files should I expect in the results folder? e.g. under `val_preds` and under `test` directory? Could you please put a tree structure of sub-files under the `results` and `test` directories?

**Comment by HuguesTHOMAS** [comment]
Q1: Did you check what was taking so much time? How long for one training epoch and how long for one validation epoch?
Potentially reducing the number of validation step can help but also verify that the validation code can handle a list of validation set instead of a single index.

Q2: CPU RAM will not make your training faster, You should watch the CPU utilisation % with htop for example. Increasing the number of worker for input dataloader can help if the CPU is the bottleneck, but it usually is not. The GPU is probably the bottleneck and you can make the best out of it by increasing the batch size (and decreasing the number of steps per epoch by the same ratio). Be careful because GPU memory consuption is not fixed, so if you go too high, the code might randomly crash from a CUDA OOM error.

Q3: I have not use this code for a while, I am not sure what are the files that should be there, but for sure you should have a ply file with predicted labels, Look at the validation function to see what is saved exactly, and use the https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/master/plot_convergence.py file to show some plots of the training and validation

**Comment by shayan-nikoo** [comment]
Thanks for your prompt reply. 

> Q1: Did you check what was taking so much time? How long for one training epoch and how long for one validation epoch? Potentially reducing the number of validation step can help but also verify that the validation code can handle a list of validation set instead of a single index.

 Each epoch takes about 11 mins and validation about 30 sec.  Do you mean `validation_szie` by num_validation step?

> Q2: CPU RAM will not make your training faster, You should watch the CPU utilisation % with htop for example. Increasing the number of worker for input dataloader can help if the CPU is the bottleneck, but it usually is not. The GPU is probably the bottleneck and you can make the best out of it by increasing the batch size (and decreasing the number of steps per epoch by the same ratio). Be careful because GPU memory consuption is not fixed, so if you go too high, the code might randomly crash from a CUDA OOM error.

So I should increase `batch_num` (now is 6) and decrease `epoch_steps` which is 500. I should find a good proportion for these. Is this what you mean?

> Q3: I have not use this code for a while, I am not sure what are the files that should be there, but for sure you should have a ply file with predicted labels, Look at the validation function to see what is saved exactly, and use the https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/master/plot_convergence.py file to show some plots of the training and validation

Sure, I got predictions.
